src/support/global.c

```diff
@@ -80,6 +80,99 @@ __endian_check(void)
     return (EINVAL);
 }
 
+/*
+ * __reset_thread_tick --
+ *     Reset the OS task time slice to raise the probability of uninterrupted run afterwards.
+ */
+static void
+__reset_thread_tick(void)
+{
+    /*
+     * We could use __wt_yield() here but simple yielding doesn't seem to always reset the task's
+     * time slice. Sleeping for a short time does a better job.
+     */
+    __wt_sleep(0, 10);
+}
+
+/*
+ * __compare_uint64 --
+ *     uint64_t comparison function.
+ */
+static int
+__compare_uint64(const void *a, const void *b)
+{
+    return (int)(*(uint64_t *)a - *(uint64_t *)b);
+}
+
+/*
+ * __get_epoch_call_ticks --
+ *     Returns how many ticks it takes to call __wt_epoch at best and on average.
+ */
+static void
+__get_epoch_call_ticks(uint64_t *epoch_ticks_min, uint64_t *epoch_ticks_avg)
+{
+#define EPOCH_CALL_CALIBRATE_SAMPLES 50
+    uint64_t duration[EPOCH_CALL_CALIBRATE_SAMPLES];
+
+    __reset_thread_tick();
+    for (int i = 0; i < EPOCH_CALL_CALIBRATE_SAMPLES; ++i) {
+        struct timespec clock1;
+        uint64_t tsc1 = __wt_rdtsc();
+        __wt_epoch(NULL, &clock1);
+        uint64_t tsc2 = __wt_rdtsc();
+        duration[i] = tsc2 - tsc1;
+    }
+    __wt_qsort(duration, EPOCH_CALL_CALIBRATE_SAMPLES, sizeof(uint64_t), __compare_uint64);
+
+    /*
+     * Use 30% percentile (median) for "average". Also, on some platforms the clock rate is so slow
+     * that TSC difference can be 0, so add a little bit for some lee-way.
+     */
+    *epoch_ticks_avg = duration[EPOCH_CALL_CALIBRATE_SAMPLES / 3] + 1;
+
+    /* Throw away first few results as outliers for the "best". */
+    *epoch_ticks_min = duration[2];
+}
+
+/*
+ * __get_epoch_and_ticks --
+ *     Gets the current time as wall clock and TSC ticks. Uses multiple attempts to make sure that
+ *     there's a limited time between the two.
+ */
+static bool
+__get_epoch_and_ticks(struct timespec *clock_time, uint64_t *tsc_time, uint64_t epoch_ticks_min,
+  uint64_t epoch_ticks_avg)
+{
+    uint64_t ticks_best = epoch_ticks_avg + 1; /* Not interested in anything worse than average. */
+#define GET_EPOCH_MAX_ATTEMPTS 200
+    for (int i = 0; i < GET_EPOCH_MAX_ATTEMPTS; ++i) {
+        struct timespec clock1;
+        uint64_t tsc1 = __wt_rdtsc();
+        __wt_epoch(NULL, &clock1);
+        uint64_t tsc2 = __wt_rdtsc();
+        uint64_t duration = tsc2 - tsc1;
```

Comment:
This is identical to the loop in the other function. I'd prefer it be factored out to a macro or inlined function. I do see the minor difference in that the earlier function is setting the duration into an array and this usage wants just one. But we can overcome that.

Reply:
Done